jetty-core/jetty-io/src/test/java/org/eclipse/jetty/io/BufferedContentSinkTest.java

```diff
@@ -0,0 +1,511 @@
+
+import org.eclipse.jetty.io.content.AsyncContent;
+import org.eclipse.jetty.io.content.BufferedContentSink;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.awaitility.Awaitility.await;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class BufferedContentSinkTest
+{
+    private ArrayByteBufferPool.Tracking _bufferPool;
+
+    @BeforeEach
+    public void setUp()
+    {
+        _bufferPool = new ArrayByteBufferPool.Tracking();
+    }
+
+    @AfterEach
+    public void tearDown()
+    {
+        assertThat("Leaks: " + _bufferPool.dumpLeaks(), _bufferPool.getLeaks().size(), is(0));
+    }
+
+    @Test
+    public void testConstructor()
+    {
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 0, 1));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, -1, 1));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 1, 0));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 1, -1));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 0, 0));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, -1, -1));
+        assertThrows(IllegalArgumentException.class, () -> new BufferedContentSink(new AsyncContent(), _bufferPool, true, 1, 2));
+    }
+
+    @Test
+    public void testWriteInvokesDemandCallback() throws Exception
+    {
+        try (AsyncContent async = new AsyncContent())
+        {
+            BufferedContentSink buffered = new BufferedContentSink(async, _bufferPool, true, 4096, 4096);
+
+            CountDownLatch latch = new CountDownLatch(1);
+            async.demand(latch::countDown);
+            assertFalse(latch.await(250, TimeUnit.MILLISECONDS));
+
+            buffered.write(true, UTF_8.encode("one"), Callback.NOOP);
+
+            assertTrue(latch.await(5, TimeUnit.SECONDS));
+
+            Content.Chunk chunk = async.read();
+            assertNotNull(chunk);
+            chunk.release();
+        }
+    }
+
```

Comment:
Please add a test for a small write followed by a large write from within the callback.

Reply:
Done.